### PR TITLE
Conversions conversion to Java

### PR DESCRIPTION
**Pull request by cfifty**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -6,11 +6,13 @@
...
 lib_managed/
 src_managed/
 project/boot/
 project/plugins/project/
 
 .idea/
-manifest.json
\ No newline at end of file
+manifest.json
+.travis.yml
```
Comment (jk2227) [inline .gitignore]: why are you adding .travis.yml to .gitignore?

File: .travis.yml [hunk 2]
```diff
@@ -4,7 +4,11 @@ scala:
 jdk:
   - oraclejdk8
 
-sudo: false
+addons:
```
Comment (jk2227) [inline .travis.yml]: why did you undo this change? the sudo: false is in there for a reason

File: Documentation/Preparator.txt [hunk 3]
```diff
@@ -0,0 +1 @@
+Prepare simply passes the ratings from TrainingData to PreparedData
```
Comment (jk2227) [inline Documentation/Preparator.txt]: instead of having a documentation folder I think it would make more sense to consolidate all the information in a markdown

File: src/main/java/org/template/recommendation/Algorithm.java [hunk 4]
```diff
@@ -128,11 +140,218 @@ public FilterCorrelators(String actionName, List<String> itemIDs) {
...
+            Optional<Map<String, JsonAST.JValue>> oRankPropMap = t._2()._2();
+
+            if (oFieldsPropMap.isPresent() && oRankPropMap.isPresent()){
+                Map<String, JsonAST.JValue> fieldPropMap = oFieldsPropMap.get();
+                Map<String, JsonAST.JValue> rankPropMap = oRankPropMap.get();
+                HashMap<String, JsonAST.JValue> newMap = new HashMap<>(fieldPropMap);
+                newMap.putAll(rankPropMap);
+                return new Tuple2<>(item, newMap);
+
+            }else if (oFieldsPropMap.isPresent()){
```
Comment (jk2227) [inline src/main/java/org/template/recommendation/Algorithm.java]: style - can we please leave a space between } and the else if?

File: src/main/java/org/template/recommendation/Algorithm.java [hunk 5]
```diff
@@ -128,11 +140,218 @@ public FilterCorrelators(String actionName, List<String> itemIDs) {
...
+            Optional<Map<String, JsonAST.JValue>> oFieldsPropMap = t._2()._1();
+            Optional<Map<String, JsonAST.JValue>> oRankPropMap = t._2()._2();
+
+            if (oFieldsPropMap.isPresent() && oRankPropMap.isPresent()){
+                Map<String, JsonAST.JValue> fieldPropMap = oFieldsPropMap.get();
+                Map<String, JsonAST.JValue> rankPropMap = oRankPropMap.get();
+                HashMap<String, JsonAST.JValue> newMap = new HashMap<>(fieldPropMap);
+                newMap.putAll(rankPropMap);
+                return new Tuple2<>(item, newMap);
+
```
Comment (jk2227) [inline src/main/java/org/template/recommendation/Algorithm.java]: unnecessary space

File: src/main/java/org/template/recommendation/Algorithm.java [hunk 6]
```diff
@@ -128,11 +140,218 @@ public FilterCorrelators(String actionName, List<String> itemIDs) {
...
+            String itemID = t._1();
+            Optional<Map<String, JsonAST.JValue>> oPropMap = t._2()._1();
+            Optional<Double> oRank = t._2()._2();
+
+            if (oPropMap.isPresent() && oRank.isPresent()){
+                Map<String, JsonAST.JValue> propMap = oPropMap.get();
+                HashMap<String, JsonAST.JValue> newMap = new HashMap<>(propMap);
+                newMap.put(fieldName, new JsonAST.JDouble(oRank.get()) );
+                return new Tuple2<>(itemID, newMap);
+            }else if (oPropMap.isPresent()){
```
Comment (jk2227) [inline src/main/java/org/template/recommendation/Algorithm.java]: style

File: src/main/java/org/template/recommendation/URModel.java [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+
+    private transient static final Logger logger = LoggerFactory.getLogger(URModel.class);
+
+    private final List<Tuple2<String, IndexedDataSet>> coocurrenceMatrices;
+    private final List<JavaPairRDD<String, Map<String,JsonAST.JValue>>> propertiesRDDs;
+    private final Map<String,String> typeMappings;
+    private final boolean nullModel;
+    private final SparkContext sc;
+
+    public URModel(
```
Comment (jk2227) [inline src/main/java/org/template/recommendation/URModel.java]: lombok this

File: src/main/java/org/template/recommendation/URModel.java [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+            final scala.collection.Iterator iter = ((JsonAST.JArray) value).values().iterator();
+            while (iter.hasNext())
+                list.add(extractJvalue(dateNames, key, (JsonAST.JValue) iter.next()));
+            return list;
+        }
+        else if (value instanceof JsonAST.JString) {
+            final String s = ((JsonAST.JString) value).s();
+            if (dateNames.contains(key)) {
+                return new DateTime(s).toDate();
+            }
```
Comment (jk2227) [inline src/main/java/org/template/recommendation/URModel.java]: } and else if on same line -- style

File: src/main/java/org/template/recommendation/indexeddataset/BiDictionaryJava.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+/**
+ * Created by Alvin Zhu on 2/27/17.
+ * Mini-wrapper for org.apache.mahout.math.indexeddataset.Bimap
+ *
+ * TODO: We skipped toSeq() function.
+ * TODO: Make BiDictionaryJava a subclass of BiMapJava
+ */
+public class BiDictionaryJava extends BiMapJava{
+
+    BiDictionary bdict;
```
Comment (jk2227) [inline src/main/java/org/template/recommendation/indexeddataset/BiDictionaryJava.java]: do we care about whether this is public private protected etc?

File: src/main/java/org/template/recommendation/indexeddataset/BiMapJava.java [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+ *
+ */
+/**
+ * Immutable Bi-directional Map.
+ * @param m Map to use for forward reference
+ * @param i optional reverse map of value to key, will create one lazily if none is provided
+ *          and is required to have no duplicate reverse mappings.
+ */
+public class BiMapJava<K,V> {
+    BiMap<K,V> bmap;
```
Comment (jk2227) [inline src/main/java/org/template/recommendation/indexeddataset/BiMapJava.java]: do we care about whether this is public private protected etc?

File: src/main/java/org/template/recommendation/indexeddataset/IndexedDatasetJava.java [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+import org.apache.spark.api.java.JavaPairRDD;
+import org.apache.predictionio.data.store.java.OptionHelper;
+
+import java.util.Optional;
+
+/**
+ * Created by Alvin Zhu on 3/1/17.
+ */
+public class IndexedDatasetJava implements IndexedDataset {
+    IndexedDatasetSpark ids;
```
Comment (jk2227) [inline src/main/java/org/template/recommendation/indexeddataset/IndexedDatasetJava.java]: do we care about whether this is public private protected etc?

File: src/main/java/org/template/recommendation/similarity/SimilarityAnalysisJava.java [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+import org.apache.mahout.math.indexeddataset.IndexedDataset;
+import org.apache.mahout.math.cf.SimilarityAnalysis;
+
+
+public class SimilarityAnalysisJava {
+
+    public static List<IndexedDataset> crossOccurrenceDownsampled(
+            List<DownsamplableCrossOccurrenceDataset> datasets,
+            int seed
+    ){
```
Comment (jk2227) [inline src/main/java/org/template/recommendation/similarity/SimilarityAnalysisJava.java]: space between ) and {

File: src/main/java/org/template/recommendation/Algorithm.java (hunk 4 above)
Comment (cfifty) [inline src/main/java/org/template/recommendation/Algorithm.java]: @jk2227 direct to @zhuyifan1996 in his pull req for everything that is not conversions.java

**Comment by cfifty** [comment]
I pulled from Alvin's pull request (he made the indexeddataset folder and pulled in my preparator changes from my preparator pull req) so no.  Will fix now

File: src/main/java/org/template/recommendation/indexeddataset/BiMapJava.java (hunk 10 above)
Comment (zhuyifan1996) [inline src/main/java/org/template/recommendation/indexeddataset/BiMapJava.java]: I think protected here is fine. This is a helper data structure for IndexedDataset, and basically won't be accessed outside the package.

File: src/main/java/org/template/recommendation/indexeddataset/IndexedDatasetJava.java (hunk 11 above)
Comment (zhuyifan1996) [inline src/main/java/org/template/recommendation/indexeddataset/IndexedDatasetJava.java]: This can be made private.

**Comment by rhenwood39** [comment]
I'm going to try to fix some style things in URModel and EsClient today.

**Comment by cfifty** [comment]
Merged into Alvin's PR which was just merged into master.  This is now outdated.
